internal/request/request.go

```diff
@@ -0,0 +1,200 @@
+	mqtt "github.com/eclipse/paho.mqtt.golang"
+	log "github.com/sirupsen/logrus"
+)
+
+type PageWriter interface {
+	Write(name string, b []byte) error
+	Flush() error
+	FlushErrors(msg []string) error
+}
+
+// RequestHandler interface allows for easy mocking during testing
+type RequestHandler interface {
+	StartHandlingRequests(mqttClient mqtt.Client, config *common.CatalogConfig, wh towerapiworker.WorkHandler)
+	//parseRequest(b []byte) (*RequestMessage, error)
+}
+
+// DefaultRequestHandler implements the 3 RequestHandler methods
+type DefaultRequestHandler struct {
+}
+
+// getRequest get data from the Receptor via Stdin
+func (drh *DefaultRequestHandler) StartHandlingRequests(mqttClient mqtt.Client, config *common.CatalogConfig, wh towerapiworker.WorkHandler) {
+	defer mqttClient.Disconnect(10)
+	sigs := make(chan os.Signal, 1)
+	shutdown := make(chan struct{})
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+	startMQTTListener(mqttClient, config, wh, shutdown)
+	done := false
+	for !done {
+		select {
+		case sig := <-sigs:
+			log.Info("Signal Received")
+			fmt.Println(sig)
+			done = true
+			close(shutdown)
+		}
+	}
+	log.Info("MQTT Client Ending")
+}
+
+func startMQTTListener(mqttClient mqtt.Client, config *common.CatalogConfig, wh towerapiworker.WorkHandler, shutdown chan struct{}) {
+	ctx := context.Background()
+	topic := "out/" + config.GUID
+	log.Infof("Subscribing to topic %s", topic)
+	counter := 0
+	fn := func(client mqtt.Client, msg mqtt.Message) {
+		log.Infof("Received a MQTT request %s", string(msg.Payload()))
+		m := common.MQTTMessage{}
+		mqttDecoder := json.NewDecoder(bytes.NewReader(msg.Payload()))
+		mqttDecoder.UseNumber()
+		err := mqttDecoder.Decode(&m)
+		if err != nil {
+			log.Errorf("Error decoding mqtt json %v", err)
+			return
+		}
+		log.Infof("Process Request %s", m.URL)
+		counter++
+		go processRequest(logger.CtxWithLoggerID(ctx, counter), m.URL, config, wh, shutdown)
+	}
+
+	if token := mqttClient.Subscribe(topic, 0, fn); token.Wait() && token.Error() != nil {
+		log.Errorf("Encountered Token Error %v", token.Error())
+	}
+}
+
+func startDispatcher(ctx context.Context, config *common.CatalogConfig, wc towerapiworker.WorkChannels, pw PageWriter, wh towerapiworker.WorkHandler) {
+	glog := logger.GetLogger(ctx)
+	done := false
+	totalCount := 0
+	finishedCount := 0
+	for !done {
+		select {
+		case j := <-wc.DispatchChannel:
+			glog.Infof("Job Input Data %v", j)
+			totalCount++
+			go startWorker(ctx, config, j, wh, wc)
+		case <-wc.Shutdown:
+			done = true
+		case page := <-wc.ResponseChannel:
+			glog.Infof("Data received on response channel %s", page.Name)
+			pw.Write(page.Name, page.Data)
+		case <-wc.FinishedChannel:
+			finishedCount++
+		default:
+			if totalCount > 0 && totalCount == finishedCount {
+				done = true
+			}
+		}
+	}
+	wc.WaitChannel <- true
+}
+
+var makeTaskOperator = func(ctx context.Context, url string) catalogtask.TaskOperator {
```

Comment:
@bzwei I think your are storing the functions in variable names so you can swap them out during testing. Could we have achieved the same thing by adding another method in the interface called make.

Reply:
@bzwei can we use a factory pattern here to create the object http://blog.ralch.com/tutorial/design-patterns/golang-factory-method/

Reply:
https://www.sohamkamani.com/golang/2018-06-20-golang-factory-patterns/

Reply:
As the second link points out there are many ways of using factory pattern in go, I am actually using it to create concrete product with different factory func. 

Using a function variable to store a default factory func is a convenient way to initialize a concrete factory. Otherwise it needs to be created in main. In a large app there may be many factories, and some could be private (mocks for testing). I feel it may be hard to for main to configure them all before the app starts.

Reply:
https://medium.com/@ankur_anand/how-to-mock-in-your-go-golang-tests-b9eee7d7c266